### ISSUE TITLE
Unescaped characters in rDNS key of JSON pretty print output

### ISSUE THREAD

**Issue by yan-foto** [drwetter/testssl.sh#1506]
- **tl;dr: `rDNS` value contains Unicode control characters which causes JSON parsing to fail.**
- **ONLY happens when pretty printing!**
- *~This is probably a bug in `dig`, which can be solved by proper sanitization.~*

---

1. uname -a
`Darwin hampers.local 18.7.0 Darwin Kernel Version 18.7.0: Thu Jan 23 06:52:12 PST 2020; root:xnu-4903.278.25~1/RELEASE_X86_64 x86_64`
   
2. testssl version from the banner: testssl.sh -b 2>/dev/null | head -4 | tail -2
```
testssl.sh       3.0 from https://testssl.sh/
    (567450e 2019-11-19 15:13:14 -- )
```

3. git log | head -1 (if running from git repo)
`b0b084dcda595ae6f4c9a23acf32183042d03816`

4. openssl version used by testssl.sh: testssl.sh -b 2>/dev/null | awk -F':' '/openssl/ { print $2}'
`Using "OpenSSL 1.0.2-chacha (1.0.2k-dev)" [~183 ciphers]`

5. steps to reproduce: testssl.sh or docker command line, if possible incl. host
```bash
# This only happens with some host names, e.g., www.beverlyhills.org
$> testssl.sh -s --jsonfile bh.json www.beverlyhills.org
# Try to parse JSON using jq (any other parser yields same error)
$> jq '.' bh.json
parse error: Invalid string: control characters from U+0000 through U+001F must be escaped at line 12, column 422
# Looking for the culprit
$> pcregrep -o "[\x00-\x1F]" bh.json | hexdump
0000000 1a 0a                                          
0000002
```
A closer look reveals that at some point in `rDNS` string a [substitute character](https://en.wikipedia.org/wiki/Substitute_character), i.e., `^Z` appears.
 
6. what exactly was happening, output is needed
See above.

7. what did you expect instead?
Proper data sanitization  before outputting JSON.

**Comment by drwetter** [comment]
Hi @yan-foto,


(partly copied from #1507 )
The cause in fact seems to be the non-printable characters: ``dig -x 192.5.3.33 +timeout=1 +tries=2 +noall +answer +short`` returns also ``beverlyhillsfiber.net\032.``, so does ``host -t PTR 192.5.3.33``.

``testssl.sh -s --jsonfile bh.json www.beverlyhills.org`` does produce valid JSON for me. The terminal output is strange though, 

```Start 2020-02-14 19:27:50        -->> 192.5.3.33:443 (www.beverlyhills.org) <<--

 rDNS (192.5.3.33):      bhfiber.net.www.bhalt.net.joinbhpd.org.beverlyhillsfiber.org.bhrent-api.beverlyhills.org.bhfiber.org.beverlyhillsfiber.net▒.www.bhgov.net.rent.beverlyhills.org.beverlyhillsarts.org.beverlyhillsfiber.net.joinbhpd.com.drivenbyinnovation.net.www.bhalt.xyz.www.bhalt.org.data.beverlyhills.org.www.beverlyhills.org.bbhfiber.net.bhgov.org.drivenbyinnovation.org.bbfiber.net.bhfiber.com.
 Service detected:       HTTP
```

We need to find a solution for that.

Cheers, Dirk

**Comment by yan-foto** [comment]
> The cause in fact seems to be the non-printable characters: `dig -x 192.5.3.33 +timeout=1 +tries=2 +noall +answer +short` returns also `beverlyhillsfiber.net\032.`, so does `host -t PTR 192.5.3.33`

That was my first intuition. Personally I prefer to use drill, but I couldn't find an explicit flag to force `testssl.sh` to use it. In any case, even though it seems to be a bug with `dig`, since the problem is reproducible, I'd like to have a solution for it here and not wait for `dig`! What do you reckon?

> Also `testssl.sh -s --jsonfile bh.json www.beverlyhills.org` does produce valid JSON for me.

Yes it does, because it does not print rDNS info at all (at least not for me on `v3.0`).

> We need to find a solution for that.

Please let me know if I can be of any help here. I was not fully satisfied with my workaround PR (#1507), but was not sure what the best approach would be.

**Comment by drwetter** [comment]
1) We need to do some input validation on DNS queries. This seems to me a security bug.
2) thus it's not only affecting JSON but also the terminal output
3) tr -dc '[:print:]' is a good measure for sanitzing non-printable characters. However the PTR records are printable characters:

```
% dig -x 192.5.3.33 +timeout=1 +tries=2 +noall +answer +short  | xxd  
00000000: 7777 772e 6265 7665 726c 7968 696c 6c73  www.beverlyhills
00000010: 2e6f 7267 2e0a 7777 772e 6268 616c 742e  .org..www.bhalt.
00000020: 6f72 672e 0a6a 6f69 6e62 6870 642e 6f72  org..joinbhpd.or
00000030: 672e 0a62 6576 6572 6c79 6869 6c6c 7366  g..beverlyhillsf
00000040: 6962 6572 2e6e 6574 5c30 3332 2e0a 6472  iber.net\032..dr
00000050: 6976 656e 6279 696e 6e6f 7661 7469 6f6e  ivenbyinnovation
00000060: 2e6f 7267 2e0a 6268 6669 6265 722e 6f72  .org..bhfiber.or
00000070: 672e 0a62 6266 6962 6572 2e6e 6574 2e0a  g..bbfiber.net..
00000080: 6262 6866 6962 6572 2e6e 6574 2e0a 7777  bbhfiber.net..ww
[...]
```
5c303332 is just ascii for \032

4) There seems to be another unrelated bug with multiple PTR records, see #1508

**Comment by drwetter** [comment]
The #PRs I issued are fixing JSON pretty output too. 

Strangely, only ``jq`` complained, ``jsonlint`` was fine

**Comment by drwetter** [comment]
Thanks again for reporting @yan-foto !
